plugins/scancode-fingerprint/src/plugin_fingerprint/fingerprint.py

```diff
@@ -0,0 +1,148 @@
+from bitarray import bitdiff
+from licensedcode.tokenize import ngrams
+import hashlib
+
+HASH_LENGTH = 128
+SHINGLE_LENGTH = 3
+
+class Simhash:
+    """
+    Fingerprint class to generate fingerprints for files used for similarity matching
+    """
+    def __init__(self):
+        self.tokens = []
+
+
+    def generate_fingerprint(self):
+        """
+        Return fingerprint as a bitarray
+        """
+        weighted_hash = self.get_weighted_hash()
+        fingerprint = self.process_weighted_hash(weighted_hash)
+
+        return fingerprint
+
+
+    def hex_digest(self):
+        """
+        Return fingerprint as a hex string
+        """
+        result = None
+
+        if len(self.tokens):
+            fingerprint_binary = self.generate_fingerprint()
+            result = binascii.hexlify(fingerprint_binary)
+        return result
+
+
+    def get_weighted_hash(self):
+        """
+        Return a weighted array from the word token list.
+        """
+        result = [0] * HASH_LENGTH
+        length = len(self.tokens) - SHINGLE_LENGTH + 1
+        shingles = ngrams(self.tokens, SHINGLE_LENGTH)
+
+        if length > 0:
+            for shingle in shingles:
+                shingle = ''.join(shingle)
+                self.process_shingles(shingle, result)
+        else:
+            self.process_shingles(''.join(self.tokens), result)
+
+        return result
+
+
+    def process_weighted_hash(self, weighted_hash):
+        """
+        Return fingerprint from a weighted array as a bitarray.
+        """
+        result = bitarray()
+
+        for item in weighted_hash:
+            if item > 0:
+                result.append(1)
+            else:
+                result.append(0)
+
+        return result
+
+    @staticmethod
+    def bitarray_from_bytes(b):
+        """
+        Return bitarray from a byte string, interpreted as machine values.
+        """
+        a = bitarray()
+        a.frombytes(b)
+
+        return a
+
+
+    def process_shingles(self, shingle, weighted_list):
+        """
+        Modify weighted list wrt to shingle
+        """
+        hash = hashlib.md5(shingle.encode()).digest()
+        result = self.bitarray_from_bytes(hash)
+
+        for idx, bit in enumerate(result):
+            if bit:
+                weighted_list[idx] += 1
+            else:
+                weighted_list[idx] -= 1
+
+        return weighted_list
+
+
+    def update(self, string):
+        """
+        Update tokens by appending new tokens
+        """
+        new_tokens = string.split()
+        self.tokens += new_tokens
+
+
+    def bitarray_from_hex(self, fingerprint_hex):
+        """
+        Return bitarray from a hex string.
+        """
+        bytes = binascii.unhexlify(fingerprint_hex)
+        result = self.bitarray_from_bytes(bytes)
+
+        return result
+
+
+    def hamming_distance(self, fingerprint1, fingerprint2):
+        """
+        Return hamming distance between two given fingerprints
```

Comment:
Can you add a bit more detail here in the docstring as to what hamming distance numbers mean what? For example, what would one expect the hamming distance to be for similar files and what would it be for completely different files. You can sort of get an idea of this looking at the test cases, but it would nice to have that information in the source code for future reference.